File: src/main/java/com/nexuscommerce/common/dto/ApiResponse.java
package com.nexuscommerce.common.dto;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.time.Instant;

/**
 * Uniform envelope for every API response in NexusCommerce.
 *
 * Success:  { "success": true,  "message": "...", "data": {...}, "timestamp": "..." }
 * Error:    { "success": false, "message": "...", "data": null,  "timestamp": "..." }
 *
 * Use the static factory methods — never call the constructor directly.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(
        boolean success,
        String message,
        T data,
        Instant timestamp
) {

    // ── Success factories ────────────────────────────────────────────────────

    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(true, "Success", data, Instant.now());
    }

    public static <T> ApiResponse<T> ok(String message, T data) {
        return new ApiResponse<>(true, message, data, Instant.now());
    }

    public static <T> ApiResponse<T> created(T data) {
        return new ApiResponse<>(true, "Created successfully", data, Instant.now());
    }

    public static <T> ApiResponse<T> created(String message, T data) {
        return new ApiResponse<>(true, message, data, Instant.now());
    }

    /** Useful for DELETE or other void-result operations */
    public static ApiResponse<Void> noContent(String message) {
        return new ApiResponse<>(true, message, null, Instant.now());
    }

    // ── Error factories ──────────────────────────────────────────────────────

    public static <T> ApiResponse<T> error(String message) {
        return new ApiResponse<>(false, message, null, Instant.now());
    }

    public static <T> ApiResponse<T> error(String message, T errorDetails) {
        return new ApiResponse<>(false, message, errorDetails, Instant.now());
    }
}

File: src/common/api-response.ts
import type { Prisma } from '@prisma/client';

/**
 * The envelope every JSON response uses: `{ success, message, data?, timestamp }`.
 * `data` is omitted when null (the Java record had @JsonInclude(NON_NULL));
 * nulls *inside* data are kept, as Jackson kept them.
 */
export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  timestamp: string;
}

function envelope<T>(success: boolean, message: string, data: T | null | undefined): ApiResponse<T> {
  const body: ApiResponse<T> = { success, message, timestamp: new Date().toISOString() };
  if (data !== null && data !== undefined) body.data = data;
  return body;
}

export const ok = <T>(data: T, message = 'Success') => envelope(true, message, data);
export const created = <T>(data: T, message = 'Created successfully') => envelope(true, message, data);
export const noContent = (message: string) => envelope<never>(true, message, null);
export const failure = <T = never>(message: string, details?: T) => envelope<T>(false, message, details);

// ── JSON number fidelity ────────────────────────────────────────────────────
// Jackson wrote BigDecimal with its scale ("price": 499.00) and longs exactly.
// JSON.rawJSON (Node ≥ 21) lets us emit the same digits instead of a lossy
// JS number.

declare global {
  interface JSON {
    rawJSON(text: string): unknown;
  }
}

/** A numeric(12,2) money value, serialized with exactly two decimals. */
export function money(value: Prisma.Decimal): unknown {
  return JSON.rawJSON(value.toFixed(2));
}

export function bigint(value: bigint | null | undefined): unknown {
  return value === null || value === undefined ? null : JSON.rawJSON(value.toString());
}
